test/lit/passes/optimize-casts.wast

```diff
@@ -387,6 +420,789 @@
+      (ref.cast $B
+        (ref.as_non_null
+          (local.get $x)
+        )
+      )
+    )
+  )
+
+  ;; CHECK:      (func $no-move-over-self-tee (type $structref_structref_=>_none) (param $x structref) (param $y structref)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.get $x)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.cast $A
+  ;; CHECK-NEXT:    (local.tee $x
+  ;; CHECK-NEXT:     (local.get $x)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $no-move-over-self-tee (param $x (ref null struct)) (param $y (ref null struct))
+    (drop
+      (local.get $x)
+    )
+    (drop
+      ;; We do not move this ref.cast of $x because $x is set by the local.tee,
+      ;; and we do not move casts past a set of a local index. This is treated
+      ;; like a local.set and we do not have a special case for this.
+      (ref.cast $A
+        (local.tee $x
+          (local.get $x)
+        )
+      )
+    )
+  )
+
+  ;; CHECK:      (func $move-over-tee (type $structref_structref_=>_none) (param $x structref) (param $y structref)
+  ;; CHECK-NEXT:  (local $a structref)
+  ;; CHECK-NEXT:  (local $3 (ref $A))
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.tee $3
+  ;; CHECK-NEXT:    (ref.cast $A
+  ;; CHECK-NEXT:     (local.get $x)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.cast $A
+  ;; CHECK-NEXT:    (local.tee $a
+  ;; CHECK-NEXT:     (local.get $3)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $move-over-tee (param $x (ref null struct)) (param $y (ref null struct))
+    (local $a (ref null struct))
+    (drop
+      (local.get $x)
+    )
+    (drop
+      ;; We can move this ref.cast because the local.tee sets another local index.
+      (ref.cast $A
+        (local.tee $a
+          (local.get $x)
+        )
+      )
+    )
+  )
+
+  ;; CHECK:      (func $move-identical-repeated-casts (type $ref|struct|_=>_none) (param $x (ref struct))
+  ;; CHECK-NEXT:  (local $1 (ref $A))
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.tee $1
+  ;; CHECK-NEXT:    (ref.cast $A
+  ;; CHECK-NEXT:     (local.get $x)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.cast $A
+  ;; CHECK-NEXT:    (local.get $1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.cast $A
+  ;; CHECK-NEXT:    (local.get $1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $move-identical-repeated-casts (param $x (ref struct))
+    (drop
+      (local.get $x)
+    )
+    (drop
+      ;; This is the first ref.cast $A which appears, so it will be moved
+      ;; up to the first local.get $x.
+      (ref.cast $A
+        (local.get $x)
+      )
+    )
+    (drop
+      ;; Since a ref.cast $A has already been moved, this will not move.
+      (ref.cast $A
```

Comment:
Both comments here seem confusing, since it's not observable which cast was moved. But the point of the test is to see that the result makes sense and this corner case doesn't cause a problem of any kind.

Reply:
If it is not visible which one is selected, should I mention which one the implementation selects?